### ISSUE TITLE
Segfault in RopePieceBTreeNode::erase

### ISSUE THREAD

**Issue by markdrayton** [iovisor/bcc#664]
This code segfaults in libbcc reproducibly:

``` python
from bcc import BPF

prog = """
#include <linux/memcontrol.h>
#include <linux/mm_types.h>

void trace(struct pt_regs *ctx, struct page *page) {
    bpf_trace_printk("%lu\\n", (unsigned long)page->mapping);
}
"""

b = BPF(text=prog)
b.attach_kprobe(event="mem_cgroup_commit_charge", fn_name="trace")
```

This is on a Ubuntu 14.04 host, BCC trunk, kernel 4.6.0, clang/LLVM 3.7 (as described in https://github.com/iovisor/bcc/blob/master/INSTALL.md).

Backtrace from GDB below. I tried `-DCMAKE_C_FLAGS=-g` but haven't been able to get the symbols working, unfortunately.

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff4ce1564 in (anonymous namespace)::RopePieceBTreeNode::erase(unsigned int, unsigned int) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
(gdb) bt
#0  0x00007ffff4ce1564 in (anonymous namespace)::RopePieceBTreeNode::erase(unsigned int, unsigned int) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#1  0x00007ffff4ce20be in clang::RewriteBuffer::ReplaceText(unsigned int, unsigned int, llvm::StringRef) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#2  0x00007ffff4ce3d56 in clang::Rewriter::ReplaceText(clang::SourceLocation, unsigned int, llvm::StringRef) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#3  0x00007ffff44221d0 in ebpf::ProbeVisitor::VisitMemberExpr(clang::MemberExpr*) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#4  0x00007ffff44688d7 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseMemberExpr(clang::MemberExpr*) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#5  0x00007ffff4460d9d in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseImplicitCastExpr(clang::ImplicitCastExpr*) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#6  0x00007ffff4466da0 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseCStyleCastExpr(clang::CStyleCastExpr*) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#7  0x00007ffff446af90 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseCallExpr(clang::CallExpr*) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#8  0x00007ffff44600c5 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseCompoundStmt(clang::CompoundStmt*) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#9  0x00007ffff4468bdf in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseFunctionHelper(clang::FunctionDecl*) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#10 0x00007ffff44278e5 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseDecl(clang::Decl*) [clone .part.6918] () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#11 0x00007ffff4427fe6 in ebpf::ProbeConsumer::HandleTopLevelDecl(clang::DeclGroupRef) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#12 0x00007ffff44c23ed in clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#13 0x00007ffff464fcf6 in clang::ParseAST(clang::Sema&, bool, bool) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#14 0x00007ffff44a0ae6 in clang::FrontendAction::Execute() () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#15 0x00007ffff447f678 in clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#16 0x00007ffff44153a2 in ebpf::ClangLoader::parse(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >*, std::unique_ptr<std::vector<ebpf::TableDesc, std::allocator<ebpf::TableDesc> >, std::default_delete<std::vector<ebpf::TableDesc, std::allocator<ebpf::TableDesc> > > >*, std::string const&, bool, char const**, int) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#17 0x00007ffff43c6b23 in ebpf::BPFModule::load_cfile(std::string const&, bool, char const**, int) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#18 0x00007ffff43cc52b in ebpf::BPFModule::load_string(std::string const&, char const**, int) () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#19 0x00007ffff43c6074 in bpf_module_create_c_from_string () from /usr/lib/x86_64-linux-gnu/libbcc.so.0
#20 0x00007ffff67e4adc in ffi_call_unix64 () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#21 0x00007ffff67e440c in ffi_call () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#22 0x00007ffff69fba0e in _ctypes_callproc () from /usr/lib/python2.7/lib-dynload/_ctypes.x86_64-linux-gnu.so
#23 0x00007ffff69fcf9e in ?? () from /usr/lib/python2.7/lib-dynload/_ctypes.x86_64-linux-gnu.so
#24 0x0000000000499be5 in PyEval_EvalFrameEx ()
#25 0x00000000004a1c9a in ?? ()
#26 0x00000000004dfe94 in ?? ()
#27 0x0000000000505f96 in PyObject_Call ()
#28 0x00000000004de41a in ?? ()
#29 0x00000000005039eb in ?? ()
#30 0x0000000000499be5 in PyEval_EvalFrameEx ()
#31 0x00000000004a1634 in ?? ()
#32 0x000000000044e4a5 in PyRun_FileExFlags ()
#33 0x000000000044ec9f in PyRun_SimpleFileExFlags ()
#34 0x000000000044f904 in Py_Main ()
#35 0x00007ffff7818ec5 in __libc_start_main (main=0x44f9c2 <main>, argc=2, argv=0x7fffffffe758, init=<optimised out>, fini=<optimised out>, rtld_fini=<optimised out>, stack_end=0x7fffffffe748)
    at libc-start.c:287
#36 0x0000000000578c4e in _start ()
```

**Comment by 4ast** [comment]
looks like clang bug, but I couldn't reproduce it in my setup with llvm trunk
could you try to isolate it by calling bpf_module_create_c_from_string() directly
from test_rewriter_bug.c ? probably will be easier to debug without python.
@drzaeus77 could you try to repro on build bots?
If bots are affected, we'd need to fix it sooner than later.

**Comment by drzaeus77** [comment]
Reproducible on buildbot with clang 3.7, same backtrace. Also reproducible on trunk clang with debug symbols. Clang gave additional assert message:

`/home/bblanco/work/llvm/tools/clang/lib/Rewrite/Rewriter.cpp:214: unsigned int clang::Rewriter::getLocationOffsetAndFileID(clang::SourceLocation, clang::FileID&) const: Assertion Loc.isValid() && "Invalid location" failed.`

```
#0  0x00007ffff738404f in raise () from /usr/lib/libc.so.6
#1  0x00007ffff738547a in abort () from /usr/lib/libc.so.6
#2  0x00007ffff737cea7 in __assert_fail_base () from /usr/lib/libc.so.6
#3  0x00007ffff737cf52 in __assert_fail () from /usr/lib/libc.so.6
#4  0x00007ffff14935da in clang::Rewriter::getLocationOffsetAndFileID (this=0x8aed70, Loc=..., FID=...) at /home/bblanco/work/llvm/tools/clang/lib/Rewrite/Rewriter.cpp:214
#5  0x00007ffff149308d in clang::Rewriter::getRangeSize (this=0x8aed70, Range=..., opts=...) at /home/bblanco/work/llvm/tools/clang/lib/Rewrite/Rewriter.cpp:132
#6  0x00007ffff14931f5 in clang::Rewriter::getRangeSize (this=0x8aed70, Range=..., opts=...) at /home/bblanco/work/llvm/tools/clang/lib/Rewrite/Rewriter.cpp:158
#7  0x00007ffff00695cd in clang::Rewriter::ReplaceText (this=0x8aed70, range=..., NewStr=...) at /home/bblanco/work/llvm/include/../tools/clang/include/clang/Rewrite/Core/Rewriter.h:144
#8  0x00007ffff005b70b in ebpf::ProbeVisitor::VisitMemberExpr (this=0xb0be50, E=0x2238360) at /home/bblanco/work/bcc/src/cc/frontends/clang/b_frontend_action.cc:245
#9  0x00007ffff017bac5 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::WalkUpFromMemberExpr (this=0xb0be50, S=0x2238360) at /home/bblanco/work/llvm/build/include/../tools/clang/include/clang/AST/StmtNodes.inc:579
#10 0x00007ffff019112f in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseMemberExpr (this=0xb0be50, S=0x2238360, Queue=0x7fffffffb0a8) at /home/bblanco/work/llvm/include/../tools/clang/include/clang/AST/RecursiveASTVisitor.h:2098
#11 0x00007ffff0177e95 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::dataTraverseNode (this=0xb0be50, S=0x2238360, Queue=0x7fffffffb0a8) at /home/bblanco/work/llvm/build/include/../tools/clang/include/clang/AST/StmtNodes.inc:579
#12 0x00007ffff016777a in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseStmt (this=0xb0be50, S=0x2238498, Queue=0x0) at /home/bblanco/work/llvm/include/../tools/clang/include/clang/AST/RecursiveASTVisitor.h:616
#13 0x00007ffff01a8457 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseFunctionHelper (this=0xb0be50, D=0x22381c0) at /home/bblanco/work/llvm/include/../tools/clang/include/clang/AST/RecursiveASTVisitor.h:1890
#14 0x00007ffff016173e in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseFunctionDecl (this=0xb0be50, D=0x22381c0) at /home/bblanco/work/llvm/include/../tools/clang/include/clang/AST/RecursiveASTVisitor.h:1895
#15 0x00007ffff0068891 in clang::RecursiveASTVisitor<ebpf::ProbeVisitor>::TraverseDecl (this=0xb0be50, D=0x22381c0) at /home/bblanco/work/llvm/build/include/../tools/clang/include/clang/AST/DeclNodes.inc:377
#16 0x00007ffff0065cd7 in ebpf::ProbeConsumer::HandleTopLevelDecl (this=0xb0be40, Group=...) at /home/bblanco/work/bcc/src/cc/frontends/clang/b_frontend_action.cc:710
#17 0x00007ffff0356311 in clang::MultiplexConsumer::HandleTopLevelDecl (this=0x10055a0, D=...) at /home/bblanco/work/llvm/tools/clang/lib/Frontend/MultiplexConsumer.cpp:278
#18 0x00007ffff0724e3a in clang::ParseAST (S=..., PrintStats=false, SkipFunctionBodies=false) at /home/bblanco/work/llvm/tools/clang/lib/Parse/ParseAST.cpp:158
#19 0x00007ffff0329ef5 in clang::ASTFrontendAction::ExecuteAction (this=0x7fffffffbf50) at /home/bblanco/work/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:558
#20 0x00007ffff032998b in clang::FrontendAction::Execute (this=0x7fffffffbf50) at /home/bblanco/work/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:458
#21 0x00007ffff02c35be in clang::CompilerInstance::ExecuteAction (this=0x7fffffffc018, Act=...) at /home/bblanco/work/llvm/tools/clang/lib/Frontend/CompilerInstance.cpp:871
#22 0x00007ffff004cf8b in ebpf::ClangLoader::parse (this=0x8268a0, mod=0x7e6b70, tables=0x7e6bb8, 
    file="\n#include <linux/memcontrol.h>\n#include <linux/mm_types.h>\n\nvoid trace(struct pt_regs *ctx, struct page *page) {\n    bpf_trace_printk(\"%lu\\n\", (unsigned long)page->mapping);\n}\n", in_memory=true, cflags=0x7ffff6ad19a0, ncflags=0)
    at /home/bblanco/work/bcc/src/cc/frontends/clang/loader.cc:227
#23 0x00007fffeffb8495 in ebpf::BPFModule::load_cfile (this=0x7e6b10, file="\n#include <linux/memcontrol.h>\n#include <linux/mm_types.h>\n\nvoid trace(struct pt_regs *ctx, struct page *page) {\n    bpf_trace_printk(\"%lu\\n\", (unsigned long)page->mapping);\n}\n", 
    in_memory=true, cflags=0x7ffff6ad19a0, ncflags=0) at /home/bblanco/work/bcc/src/cc/bpf_module.cc:325
#24 0x00007fffeffbae57 in ebpf::BPFModule::load_string (this=0x7e6b10, text="\n#include <linux/memcontrol.h>\n#include <linux/mm_types.h>\n\nvoid trace(struct pt_regs *ctx, struct page *page) {\n    bpf_trace_printk(\"%lu\\n\", (unsigned long)page->mapping);\n}\n", 
    cflags=0x7ffff6ad19a0, ncflags=0) at /home/bblanco/work/bcc/src/cc/bpf_module.cc:730
#25 0x00007fffeffb459d in bpf_module_create_c_from_string (text=0x7ffff6be4a70 "\n#include <linux/memcontrol.h>\n#include <linux/mm_types.h>\n\nvoid trace(struct pt_regs *ctx, struct page *page) {\n    bpf_trace_printk(\"%lu\\n\", (unsigned long)page->mapping);\n}\n", 
    flags=0, cflags=0x7ffff6ad19a0, ncflags=0) at /home/bblanco/work/bcc/src/cc/bpf_common.cc:40
#26 0x00007ffff5f701c8 in ffi_call_unix64 () from /usr/lib/libffi.so.6
#27 0x00007ffff5f6fc2a in ffi_call () from /usr/lib/libffi.so.6
```

**Comment by 4ast** [comment]
spoke too soon. The latest clang/llvm trunk doesn't even build for me:
```
/w/bcc/src/cc/bpf_module.cc: In member function ‘int ebpf::BPFModule::run_pass_manager(llvm::Module&)’:
/w/bcc/src/cc/bpf_module.cc:399:34: error: ‘createAlwaysInlinerPass’ was not declared in this scope
```

**Comment by markdrayton** [comment]
Here's a minimal repro without Python:

```c
#include "bpf_common.h"

int main(int argc, char **argv) {
  const char *str =
    "#include <linux/mm_types.h>\n"
    "unsigned long trace(struct pt_regs *ctx, struct page *page) {\n"
    "    return (unsigned long)page->mapping;\n"
    "}\n";

  void *mod = bpf_module_create_c_from_string(str, 4, NULL, 0);
  return !(mod != NULL);
}
```

**Comment by drzaeus77** [comment]
Update:

I narrowed this failure down to the clang::MemberExpr object being created with an empty OperatorLoc during the semantic analysis phase. The rewriter depends on a valid OperatorLoc in order to rewrite the input text properly. I have tried a small patch to llvm that fills in the field, but I fear that our dependence on the rewriter approach is too fragile to survive much longer, especially now that we see what guarantees the AST objects do/don't give to us.

I'm planning to look into an alternative implementation that works on LLVM IR rather than AST. In the meantime, I'll push a commit that gives an error message when op is invalid rather than crashing.
